File: frontend/src/lib/mastery.js
export const MASTERY_CONFIG = {
  excellent: {
    label: 'Excellent',
    color: 'green',
    bg: 'bg-green-100',
    text: 'text-green-800',
    bar: 'bg-green-500',
    border: 'border-green-200',
  },
  strong: {
    label: 'Strong',
    color: 'green',
    bg: 'bg-green-50',
    text: 'text-green-700',
    bar: 'bg-green-400',
    border: 'border-green-200',
  },
  improving: {
    label: 'Improving',
    color: 'yellow',
    bg: 'bg-yellow-100',
    text: 'text-yellow-800',
    bar: 'bg-yellow-500',
    border: 'border-yellow-200',
  },
  weak: {
    label: 'Weak',
    color: 'orange',
    bg: 'bg-orange-100',
    text: 'text-orange-800',
    bar: 'bg-orange-500',
    border: 'border-orange-200',
  },
  critical: {
    label: 'Critical',
    color: 'red',
    bg: 'bg-red-100',
    text: 'text-red-800',
    bar: 'bg-red-500',
    border: 'border-red-200',
  },
  not_attempted: {
    label: 'Not Attempted',
    color: 'gray',
    bg: 'bg-gray-100',
    text: 'text-gray-600',
    bar: 'bg-gray-300',
    border: 'border-gray-200',
  },
}

export const CONFIDENCE_LABELS = {
  not_attempted: 'N/A',
  low: 'Low',
  medium: 'Medium',
  high: 'High',
}

export function masteryStyle(level) {
  return MASTERY_CONFIG[level] || MASTERY_CONFIG.not_attempted
}

File: backend/app/services/mastery.py
"""Shared mastery and confidence calculations for weakness map and practice."""

MASTERY_EXCELLENT = "excellent"
MASTERY_STRONG = "strong"
MASTERY_IMPROVING = "improving"
MASTERY_WEAK = "weak"
MASTERY_CRITICAL = "critical"
MASTERY_NOT_ATTEMPTED = "not_attempted"

CONFIDENCE_NOT_ATTEMPTED = "not_attempted"
CONFIDENCE_LOW = "low"
CONFIDENCE_MEDIUM = "medium"
CONFIDENCE_HIGH = "high"

MASTERED_LEVELS = {MASTERY_EXCELLENT, MASTERY_STRONG}
WEAK_LEVELS = {MASTERY_WEAK, MASTERY_CRITICAL}


def mastery_level_for(accuracy: float, questions_attempted: int) -> str:
    if questions_attempted == 0:
        return MASTERY_NOT_ATTEMPTED
    if accuracy >= 90:
        return MASTERY_EXCELLENT
    if accuracy >= 75:
        return MASTERY_STRONG
    if accuracy >= 60:
        return MASTERY_IMPROVING
    if accuracy >= 40:
        return MASTERY_WEAK
    return MASTERY_CRITICAL


def mastery_label(level: str) -> str:
    labels = {
        MASTERY_EXCELLENT: "Excellent",
        MASTERY_STRONG: "Strong",
        MASTERY_IMPROVING: "Improving",
        MASTERY_WEAK: "Weak",
        MASTERY_CRITICAL: "Critical",
        MASTERY_NOT_ATTEMPTED: "Not Attempted",
    }
    return labels.get(level, level)


def confidence_level_for(questions_attempted: int) -> str:
    if questions_attempted == 0:
        return CONFIDENCE_NOT_ATTEMPTED
    if questions_attempted < 4:
        return CONFIDENCE_LOW
    if questions_attempted < 11:
        return CONFIDENCE_MEDIUM
    return CONFIDENCE_HIGH


def weakness_level_for(accuracy: float) -> str:
    """Legacy storage field on topic_performance; maps to mastery tier."""
    return mastery_level_for(accuracy, questions_attempted=1)
